File: Design-Patterns-Using-JavaScript/Template/TemplatePatternData.js
// Define AddDataTemplate
class AddDataTemplate {
    addData() {
        this.open();
        this.add();
        this.close();
    }

    open() {
        throw new Error("open method must be implemented");
    }

    add() {
        throw new Error("add method must be implemented");
    }

    close() {
        throw new Error("close method must be implemented");
    }
}

// Define AddDataToFile extending AddDataTemplate
class AddDataToFile extends AddDataTemplate {
    open() {
        console.log("Open file.");
    }

    add() {
        console.log("Add data to file.");
    }

    close() {
        console.log("Close file");
    }
}

// Define AddDataToDB extending AddDataTemplate
class AddDataToDB extends AddDataTemplate {
    open() {
        console.log("Open Database.");
    }

    add() {
        console.log("Add data to Database.");
    }

    close() {
        console.log("Close Database.");
    }
}

// Client code
const addDataToDB = new AddDataToDB();
addDataToDB.addData();

/*
Open Database.
Add data to Database.
Close Database.
*/

File: Design-Patterns-Using-Ruby/Template/TemplatePatternData.rb
# Define AddDataTemplate
class AddDataTemplate
    def add_data
        open
        add
        close
    end

    def open
        raise NotImplementedError, 'Subclasses must implement the open method'
    end

    def add
        raise NotImplementedError, 'Subclasses must implement the add method'
    end

    def close
        raise NotImplementedError, 'Subclasses must implement the close method'
    end
end

# Define AddDataToFile extending AddDataTemplate
class AddDataToFile < AddDataTemplate
    def open
        puts 'Open file.'
    end

    def add
        puts 'Add data to file.'
    end

    def close
        puts 'Close file.'
    end
end

# Define AddDataToDB extending AddDataTemplate
class AddDataToDB < AddDataTemplate
    def open
        puts 'Open Database.'
    end

    def add
        puts 'Add data to Database.'
    end

    def close
        puts 'Close Database.'
    end
end

# Client code
add_data_to_db = AddDataToDB.new
add_data_to_db.add_data

=begin 
Open Database.
Add data to Database.
Close Database.
=end